src/mlpack/methods/ne/genome.hpp

```diff
@@ -0,0 +1,304 @@
+                                 std::end(depths)));
+    return (maxInputDepth + 1);
+  }
+
+  // Calculate Genome depth.
+  // It is the max depth of all output neuron genes.
+  unsigned int GenomeDepth() {
+    unsigned int numNeuron = NumNeuron();
+
+    // If empty genome.
+    if (numNeuron == 0) {
+      aDepth = 0;
+      return aDepth;
+    }
+
+    // If no hidden neuron, depth is 1.
+    if (aNumInput + aNumOutput == numNeuron) {
+      aDepth = 1;
+      return aDepth;
+    }
+    
+    // Find all OUTPUT neuron id.
+    std::vector<unsigned int> outputNeuronsId;
+    for (unsigned int i=0; i<NumNeuron(); ++i) {
+      if (aNeuronGenes[i].Type() == OUTPUT) {
+        outputNeuronsId.push_back(aNeuronGenes[i].Id());
+      }
+    }
+
+    // Get max depth of all output neurons.
+    unsigned int genomeDepth = 0;
+    for (unsigned int i=0; i<outputNeuronsId.size(); ++i) {
+      unsigned int outputNeuronDepth = NeuronDepth(outputNeuronsId[i]);
+      if (outputNeuronDepth > genomeDepth) {
+        genomeDepth = outputNeuronDepth;
+      }
+    }
+    aDepth = genomeDepth;
+
+    return aDepth;
+  }
+
+  // Set neurons' input and output to zero.
+  void Flush() {
+    for (unsigned int i=0; i<aNeuronGenes.size(); ++i) {
+      aNeuronGenes[i].aActivation = 0;
+      aNeuronGenes[i].aInput = 0;
+    }
+  }
+
+  // Activate genome. The last dimension of input is always 1 for bias. 0 means no bias.
+  void Activate(std::vector<double>& input) {
+    assert(input.size() == aNumInput);
+    Flush();
+
+    // Set inputs.
+    for (unsigned int i=0; i<aNumInput; ++i) {
+      aNeuronGenes[i].aActivation = input[i];  // assume INPUT, BIAS, OUTPUT, HIDDEN sequence
+    }
+
+    // Construct neuron id: index dictionary.
+    std::map<unsigned int, unsigned int> neuronIdToIndex;
+    for (unsigned int i=0; i<NumNeuron(); ++i) {
+      neuronIdToIndex.insert(std::pair<unsigned int, unsigned int>(aNeuronGenes[i].Id(), i));
+    }
+
+    // Activate layer by layer.
+    for (unsigned int i=0; i<aDepth; ++i) {
+      // Loop links to calculate neurons' input sum.
+      for (unsigned int j=0; j<aLinkGenes.size(); ++j) {
+        aNeuronGenes[neuronIdToIndex.at(aLinkGenes[j].ToNeuronId())].aInput +=
+          aLinkGenes[j].Weight() * aNeuronGenes[neuronIdToIndex.at(aLinkGenes[j].FromNeuronId())].aActivation;
+      }
+
+      // Loop neurons to calculate neurons' activation.
+      for (unsigned in j=aNumInput; j<aNeuronGenes.size(); ++j) {
+        double x = aNeuronGenes[j].aInput;  // TODO: consider bias. Difference?
+        aNeuronGenes[j].aInput = 0;
+
+        double y = 0;
+        switch (aNeuronGenes[j].Type()) { // TODO: revise the implementation.
+          case SIGMOID:                   // TODO: more cases.
+            y = sigmoid(x);
+            break;
+          case RELU:
+            y = relu(x);
+            break;
+          default:
+            y = sigmoid(x);
```

Comment:
I don't think it's a good idea, to reimplement the activation function or to keep a copy in the ne module. For a couple of reasons:

- it would increase maintenance
- if there is a bug in one of the activation functions we probably have to fix the same activation function twice
- we have to rewrite the tests that are already existing for all ann activation functions
- if we implement a non-existing activation function for one of the evolution methods, we can also use the function for the ann modules

Also, since we have this neat update policy, we don't go and change API's between two versions. So if we change something there should be enough time, for adaptation.